### PR TITLE
Fix matching different guest products with a single Unlimited Virtualization subscription

### PR DESCRIPTION
**Pull request by hustodemon**
See the `Fix: Allow matching any guest products for Unlimited Virtualization
subscriptions` commit for more details. I also wrote new test scenarios (42 and 43) testing this fix.

Moreover, the PR contains additional fixes and improvements:
- expose the log level to the command line - e.g. `-v/--log-level DEBUG` (this will we used by SUSE Manager: https://github.com/uyuni-project/uyuni/pull/2498)
- if log level is `DEBUG` or greater, log the whole assignment as JSON in `output-all.json`
- fixed some older test data (mostly the README.md files)
- moved SwapMove classes to correct packages
- fixed duplicate log lines when running tests
- logging test scenario number when running tests
- in the subscriptions CSV output, print the active subscriptions first (it's speeds up analysis of the data - the user doesn't need to sort the rows first)

Review *commit-by-commit* please, commit messages are descriptive.

### CODE REVIEW

**Comment by hustodemon** [comment]
@mcalmer I'd like to hear your opinion on the idea of my fix. The essence of the fix is:
An unlimited virtualization assigned to a host now covers all products installed on the guests that are "coverable" by the subscription. So, if a guest 1 has SLE 11 and guest 2 has SLE 12, they'll be both covered.

The only catch I could think of would be, if the subscription covered 2 products, that could be installed on the guest simultaneously, e.g.:
```
       HOST
        |          
      GUEST1
      - SLE 11
      - some other product coverable by the same subscription
```

With my patch, both of these products would be covered by a single UV subscription. I'm not sure if this can happen.

**Comment by mcalmer** [comment]
@hustodemon yes this happens with all the modules you need for SLES, but counting happens based on the "host" - VMs are "unlimited" and this means every VM with one or more of the listed products in the subscription are "OK".
Important is, that the CPU counting happens based on the host CPUs. The VM is not important here.

File: src/test/resources/scenario42/output.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
+{
+    "timestamp": "2015-05-01T00:00:00.000+02",
+    "matches": [
+        {
+          "system_id": 100,
+          "subscription_id": 1,
+          "product_id": 1000,
+          "cents": 50,
```
Comment (mcalmer) [inline src/test/resources/scenario42/output.json:8]: why "50" cents? What would happen in real live where a subscription covers 100 or more product ids?

File: src/test/resources/scenario43/output.json [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        "cents": 50,
+        "confirmed": true
+      },
+      {
+        "system_id": 100,
+        "subscription_id": 1,
+        "product_id": 1625,
+        "cents": 100,
+        "confirmed": false
+      },
```
Comment (mcalmer) [inline src/test/resources/scenario43/output.json:31]: 2 times the same tripple of system_id, subscription_id and product_id but with different cents. Why?
I think 1 item with 100 cent would be correct here.

File: src/test/resources/scenario43/output.json [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        "cents": 50,
+        "confirmed": true
+      },
+      {
+        "system_id": 100,
+        "subscription_id": 1,
+        "product_id": 1878,
+        "cents": 100,
+        "confirmed": false
+      },
```
Comment (mcalmer) [inline src/test/resources/scenario43/output.json:45]: same as above

File: src/test/resources/scenario43/output.json [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        "cents": 0,
+        "confirmed": false
+      },
+      {
+        "system_id": 101,
+        "subscription_id": 1,
+        "product_id": 1625,
+        "cents": 0,
+        "confirmed": true
+      },
```
Comment (mcalmer) [inline src/test/resources/scenario43/output.json:73]: Two times one with confirmed true the other with false? Why?

File: src/test/resources/scenario43/output.json [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        "product_id": 1878,
+        "cents": 0,
+        "confirmed": false
+      },
+      {
+        "system_id": 102,
+        "subscription_id": 1,
+        "product_id": 1878,
+        "cents": 0,
+        "confirmed": true
```
Comment (mcalmer) [inline src/test/resources/scenario43/output.json:100]: same as above

File: src/test/resources/scenario42/output.json (hunk 1 above)
Comment (hustodemon) [reply src/test/resources/scenario42/output.json:8]: But if you look at the next match, you'll see that the matcher uses 50 cents for the other product (`1001`). It total matching both products with subscription `1` consumes 100 cents.

Comment (hustodemon) [reply src/test/resources/scenario42/output.json:8]: (The thing is that a match can't contain 2 product ids, so it's split into 2 matches that have the same group id (not visible here) that must be both either confirmed or unconfirmed.)

File: src/test/resources/scenario43/output.json (hunk 2 above)
Comment (hustodemon) [reply src/test/resources/scenario43/output.json:31]: Similar as in the case above. These matches are part of different matching scenarios:
- the match with `100` cents would be used if **only** the product `1625` was assigned to the host
- the pair of matches with `50` cents is used if both `1625` and `1878` are assigned to the host.

File: src/test/resources/scenario43/output.json (hunk 4 above)
Comment (hustodemon) [reply src/test/resources/scenario43/output.json:73]: I think these matches have different `groupId`. One match belongs to the scenario, where subscription `1` covers both `1625` and `1299`, the other match belongs to scenarion, where `1` covers only `1625`.

I'll verify this now. I also think it'd make sense to include `groupId` ids in the output.

File: src/test/resources/scenario43/output.json (hunk 2 above)
Comment (hustodemon) [reply src/test/resources/scenario43/output.json:31]: Here the situation is complicated with the inherited virtualization subscription, but the result is correct. There are 2 situations, which both produce their own sets of matches in the output and these are same on the first glance.

The confirmed match comes from the situation, where the subscription `1` (unlimited virt.) covers both `1625` and `1878` on the guests and together with this, the (inherited virt.) subscription `2` covers the guests as well.

The unconfirmed match comes from the situation, where the subscription `1`  covers both `1625` and `1878` on the guests, BUT the subscription `2`is not confirmed. Btw, the number of the `cents` is irrelevant for unconfirmed matches, maybe we can talk about it.

**Comment by hustodemon** [comment]
After some further thinking I think there are 2 ways to improve the clarity of the `output.json`:
1. do not report the unconfirmed matches in the output: in many cases, their `cents` value is incorrect; in some cases, multiple matches, that look the same (but are not, because they have different cent group) are reported, which is confusing
1. include the cent groups in the `output.json` too: each reported match would contain a `centGroupId` field; we also might to report additional array of cent groups in the output

Just to recap, a cent group is a mechanism that I've added, so that multiple confirmed matches can consume N cents (the motivation was 1-2 subscriptions: confirming **either** 1 **or** 2 matches of a 1-2 subscription always consumes 100 cents). More details on this are in our deprecated gitlab repo (merge request 29). Ping me for any further info.

@mcalmer, @moio : I'd be happy, if we could talk about this next week or later.

**Comment by moio** [comment]
Are there downsides you can see in evicting unconfirmed matches from the output?

Do we use them anywhere or could we treat them as an internal implementation detail of the matcher?

Would testing/debugging become significantly more complex by removing them from output.json?

**Comment by hustodemon** [comment]
> Are there downsides you can see in evicting unconfirmed matches from the output?
You guessed one of the downsides in your 3rd point: we're losing some information, that _might_ be useful during debugging. At the same time, having them in the output can bring some confusion, as in some cases (as I mentioned in my latest comments), the `cents` field contains gibberish. We could either omit the `cents` field from unconfirmed matches, or we could make the cent groups part of the output, as described above.

> Do we use them anywhere or could we treat them as an internal implementation detail of the matcher?
Afaik, they are not used by the matcher output consumers (SUMA UI).

**Comment by moio** [comment]
> You guessed one of the downsides in your 3rd point: we're losing some information, that might be useful during debugging. 

Would it be an option to have this information that's only relevant to debugging in logs, thereby cleaning up output.json for its intended audience? (maybe as simple as dumping the json there)

**Comment by hustodemon** [comment]
I find this to be a nice idea! Maybe we could expose the loglevel control
and this dump could be controlled via either a cli option or by `debug`
field in /etc/rhn.conf).

Silvio Moioli <notifications@github.com> schrieb am Fr., 14. Aug. 2020,
17:14:

**Comment by mcalmer** [comment]
> I find this to be a nice idea! Maybe we could expose the loglevel control and this dump could be controlled via either a cli option or by `debug` field in /etc/rhn.conf).

We do something like this already for the Gatherer. So it should be easy to have something like this for Matcher too.

What about a mumble session to discuss this issue?

**Comment by moio** [comment]
If you need my input just ping in chat

**Comment by hustodemon** [comment]
@moio @mcalmer I improved the logging (passing log level from the CLI, logging all facts in the assignment as JSON). I played around with sorting facts in the Construction Heuristics with mixed results - I still don't know, if this is the right direction. I think the best way here would be to employ the Optaplanner benchmark tool to find the "right" parameters for big setups.

Originally, the user reported 2 problems:
- an UNLIMITED virtualization didn't cover various products installed on guests (fixed by `Fix: Allow matching any guest products for Unlimited Virtualization subscriptions`)
- the SUSE Manager subscription was unmatched (fixed by `Increase the unimproved count limit `)

The user reported additional problem, but it seems this is unrelated to these 2 (and it's very likely not a matcher bug (the CPU count for some systems in his setup is misreported, thus leading to wasted subs)).

This PR is already big and fixes 2 problems, so I would recommend merging it after review. We can address possible additional issues separately. Please re-read the PR description and review.

**Comment by mcalmer** [comment]
@moio @hustodemon In the early days we had also a list of systems not getting a subscription. This was removed long ago.
So I think the UI calculate it. The UI send a list of systems and get the matched systems back. Every system which is not in matched, miss a subscription and is an "error".
Does the UI show it? If yes, I am fine.

**Review by mcalmer — APPROVED** [review]
I can understand the unit test output again. Thanks

**Comment by hustodemon** [comment]
Fair point. We display a list of unmatched products. I need to take a look
how these are computed in SUMA. Thanks for bringing this up!

Michael Calmer <notifications@github.com> schrieb am Mi., 26. Aug. 2020,
17:20:

**Comment by hustodemon** [comment]
@mcalmer we are safe with regards to your last comment - unmatched products are computed in the way you described (we don't rely on "unconfirmed" matches in the output, so we're good to go!

**Pull request merged**